### PR TITLE
NIFI-6778 - Added rules engine service api and Easy Rules Implementation

### PR DESCRIPTION
**Pull request by YolandaMDavis**
#### Description of PR

This PR introduces Rules Engine functionality as a Controller Service with Easy Rules engine serving as the default engine.  To test functionality a Custom Processor (e.g. with InvokedScriptedProcessor) can be used to invoked the RulesEngineService along with a custom rules files.  The addtionalDetails section in the EasyRulesEngineService provides information on how to construct rules and how the service can be used by processors.

This PR includes an example Flow Template and rules file which can be used as a basis for testing interaction (see attached zip file).  The Template uses the rules engine to determine the specific relationship a  flowfile should be routed on based on the the size of the file (template uses GenerateFlowFile processors). 

[test_rules.zip](https://github.com/apache/nifi/files/3741008/test_rules.zip)

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [X] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [X] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/java/org/apache/nifi/rules/engine/EasyRulesEngineService.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+ * @see RulesEngineService
+ */
+@CapabilityDescription("Defines and execute the rules stored in NiFi or EasyRules file formats for a given set of facts. Supports " +
+        "rules stored as JSON or YAML.")
+@Tags({ "rules","rules-engine","engine","actions","facts" })
+public class EasyRulesEngineService  extends AbstractControllerService implements RulesEngineService {
+
+    static final AllowableValue YAML = new AllowableValue("YAML", "YAML", "YAML file configuration type.");
+    static final AllowableValue JSON = new AllowableValue("JSON", "JSON", "JSON file configuration type.");
+    static final AllowableValue NIFI = new AllowableValue("NIFI", "NiFi", "NIFI rules formatted file.");
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/java/org/apache/nifi/rules/engine/EasyRulesEngineService.java]: Nitpick, the display name is not all caps but the value and description are. In the generated doc, the default value is `NIFI` but in the list of options it says `NiFi`.

File: nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/java/org/apache/nifi/rules/engine/EasyRulesEngineService.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    static final AllowableValue MVEL = new AllowableValue("MVEL", "Easy Rules MVEL", "Easy Rules File format using MVFLEX Expression Language");
+    static final AllowableValue SPEL = new AllowableValue("SPEL", "Easy Rules SpEL", "Easy Rules File format using Spring Expression Language");
+
+    static final PropertyDescriptor RULES_FILE_PATH = new PropertyDescriptor.Builder()
+            .name("rules-file-path")
+            .displayName("Rules File Path")
+            .description("Rules File Location")
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/java/org/apache/nifi/rules/engine/EasyRulesEngineService.java]: Nitpick, maybe some more doc for the description ("Path to the location of the rules file" or something?)

File: nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/java/org/apache/nifi/rules/engine/EasyRulesEngineService.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            .description("File type for rules definition. Supported file types are YAML and JSON")
+            .required(true)
+            .allowableValues(JSON,YAML)
+            .defaultValue(JSON.getValue())
+            .build();
+
+    static final PropertyDescriptor RULES_FILE_FORMAT = new PropertyDescriptor.Builder()
+            .name("rules-file-format")
+            .displayName("Rules File Format")
+            .description("File format for rules. Supported formats are NiFi Rules.")
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/java/org/apache/nifi/rules/engine/EasyRulesEngineService.java]: We should include all the currently supported formats in the description

File: nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/resources/docs/org.apache.nifi.rules.engine.EasyRulesEngineService/additionalDetails.html [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+<p>
+    NiFi Rules Format consist of the following elements:
+</p>
+<table>
+    <tr><td>Name</td><td>The name of the rule.</td></tr>
+    <tr><td>Description</td><td>A description of the rule.</td></tr>
+    <tr><td>Priority</td><td>Unique number for order in which rule should be executed in the rules engine.</td></tr>
+    <tr><td>Condition</td><td>The condition that will be checked given a set of fact data. For NiFi Format the MVEL expression language is supported.</td></tr>
+    <tr><td>Action(s)</td><td>The required actions that should be performed if the condition checked returns true.  This action includes the type of action and any attributes required to execute the action.</td></tr>
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/resources/docs/org.apache.nifi.rules.engine.EasyRulesEngineService/additionalDetails.html]: Should we include some example actions here, such as EXPRESSION, ROUTE, SEND, LOG, etc.?

File: nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/java/org/apache/nifi/rules/engine/EasyRulesEngineService.java (hunk 3 above)
Comment (YolandaMDavis) [reply nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/java/org/apache/nifi/rules/engine/EasyRulesEngineService.java]: Good catch here, I forgot to update (when I only had NiFi rules format implemented:))

File: nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/resources/docs/org.apache.nifi.rules.engine.EasyRulesEngineService/additionalDetails.html (hunk 4 above)
Comment (YolandaMDavis) [reply nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/resources/docs/org.apache.nifi.rules.engine.EasyRulesEngineService/additionalDetails.html]: Yep that makes sense.  I will add those in

**Comment by YolandaMDavis** [comment]
@mattyb149 thanks for the feedback! I will add those changes in

File: nifi-assembly/pom.xml [hunk 5]
```diff
@@ -721,7 +721,12 @@ language governing permissions and limitations under the License. -->
             <version>1.10.0-SNAPSHOT</version>
             <type>nar</type>
         </dependency>
-
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-easyrules-nar</artifactId>
```
Comment (joewitt) [inline nifi-assembly/pom.xml]: not clear on size of this nar but generally we cannot add nars....unless really small

Comment (YolandaMDavis) [reply nifi-assembly/pom.xml]: good point @joewitt I will provide a profile to include this nar only if enabled for build.

File: nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/resources/docs/org.apache.nifi.rules.engine.EasyRulesEngineService/additionalDetails.html (hunk 4 above)
Comment (YolandaMDavis) [reply nifi-nar-bundles/nifi-easyrules-bundle/nifi-easyrules-service/src/main/resources/docs/org.apache.nifi.rules.engine.EasyRulesEngineService/additionalDetails.html]: @mattyb149 I may lean here more on describing the way people can define actions which can be leveraged by their own handlers

File: nifi-assembly/LICENSE [hunk 6]
```diff
@@ -2834,3 +2834,25 @@ The binary distribution of this product bundles 'JUnit Platform' and 'JUnit Jupi
   notice.
 
   You may add additional accurate notices of copyright ownership.
+
```
Comment (apiri) [inline nifi-assembly/LICENSE]: Looks like MVEL also bundles (?) ASM as per https://github.com/mvel/mvel/blob/master/ASM-LICENSE.txt.  Not sure if that is being consumed imported, but wanted to flag it.

Comment (YolandaMDavis) [reply nifi-assembly/LICENSE]: The MVEL library isn't used directly in this tooling, however it is included in easy rules library, via the easy-rules-mvel dependency and it does use Easy Rules MVELCondition objects.  Will this dependency still need to be cited?

Comment (mattyb149) [reply nifi-assembly/LICENSE]: I don't see it in the dependency tree:

```|  +- org.jeasy:easy-rules-core:jar:3.3.0:compile
[INFO] |  +- org.jeasy:easy-rules-mvel:jar:3.3.0:compile
[INFO] |  |  +- org.jeasy:easy-rules-support:jar:3.3.0:compile
[INFO] |  |  \- org.mvel:mvel2:jar:2.4.3.Final:compile
[INFO] |  \- org.jeasy:easy-rules-spel:jar:3.3.0:compile
[INFO] |     +- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:jar:2.9.8:compile
[INFO] |     |  \- com.fasterxml.jackson.core:jackson-core:jar:2.9.8:compile
[INFO] |     +- com.fasterxml.jackson.core:jackson-databind:jar:2.9.8:compile
[INFO] |     |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.9.0:compile
[INFO] |     \- org.springframework:spring-expression:jar:4.3.23.RELEASE:compile
[INFO] |        \- org.springframework:spring-core:jar:4.3.23.RELEASE:compile
[INFO] |           \- commons-logging:commons-logging:jar:1.2:compile
[INFO]```

Comment (YolandaMDavis) [reply nifi-assembly/LICENSE]: @mattyb149 chatted with @apiri offline. It looks like MVEL is a transitive dependency so I just added license info for it

**Comment by YolandaMDavis** [comment]
@mattyb149 FYI in addition to the license info update (for MVEL) I've also switched from gson to using jackson json processor.

**Comment by mattyb149** [comment]
+1 LGTM, verified the latest changes (including that the NAR is not included in the assembly by default). Thanks for this new feature! Merging to master

**Pull request closed without merging**